x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java

```diff
@@ -107,7 +114,11 @@ public static class Builder extends FieldMapper.Builder {
             mapper -> ((SemanticTextFieldType) mapper.fieldType()).modelSettings,
             XContentBuilder::field,
             Objects::toString
-        ).acceptsNull().setMergeValidator(SemanticTextFieldMapper::canMergeModelSettings);
+        ).acceptsNull()
+            .setMergeValidator(
+                ((previous, current, conflicts) -> SemanticTextFieldMapper.canMergeModelSettings(previous, current, true, conflicts))
+            )
+            .setValueMerger(SemanticTextFieldMapper::mergeModelSettings);
```

Comment:
This solves a problem I didn't expect to run into. It turns out our concept of "merging" parameter values is more like overwriting. However, for `semantic_text` fields, we want to keep the previous model settings if the new model settings are null. 

This handles the case where we have ingested some data into our `semantic_text` field, setting the model settings, and then the user uses the update mapping API to change `search_inference_id`:

```json
PUT my-index/_mapping
{
  "properties": {
    "inference_field": {
      "type": "semantic_text",
      "inference_id": "my-elser-endpoint-1",
      "search_inference_id": "my-new-endpoint"
    }
  }
}
```

In this request, the model settings are null, but this should be fine because we should be able to use the previous model settings and merge them into the updated mapping to get the new complete mapping.

I added the concept of a "value merger" to field mapping parameters to accomplish this. By default, the merge is performed by always taking the new value (i.e. the pre-existing behavior), but developers can now define custom value merging behavior as necessary.

Reply:
> It turns out our concept of "merging" parameter values is more like overwriting.

Indeed, when updating a field definition, users are required to provide the entire definition to avoid ambiguity.  
Although, our use of a parameter to expose `model_settings` is more of a workaround since no better alternatives exist.  
It would be better to avoid introducing a general concept for an unsupported use case from the start.

How about modifying `SemanticTextFieldMapper#getMergeBuilder` like this?
```java
@Override
public FieldMapper.Builder getMergeBuilder() {
    var builder = new SemanticTextFieldMapper.Builder(leafName(), fieldType().indexVersionCreated, fieldType().getChunksField().bitsetProducer());
    builder.init(this);
    builder.setModelSettings(fieldType().modelSettings);
    return builder;
}
```

This way, we set the current model settings as the default when merging mappings. If the new definition overrides it, the merge will fail; otherwise, the existing model_settings is used seamlessly. Since updating an existing `model_settings` to `null` is already prohibited, this approach would ensure the model settings are sealed once defined.

Reply:
Thanks for the guidance, this approach looks much better :)

Reply:
I tried this approach and unfortunately it doesn't work :(

`getMergeBuilder` builds the previous `FieldMapper` in the `previous` : `current` comparison. Additionally, the model settings are already set in `previous` through the `init(this)` call in `getMergeBuilder()`.

What we need is a way to modify `current` before `canMergeModelSettings` is called to set the model settings appropriately. Any hooks we could use to do that?

Reply:
I think we may be able to handle this by modifying the current `FieldMapper` (`mergeWith`) in [SemanticTextFieldMapper.Builder#merge](https://github.com/elastic/elasticsearch/blob/d439a8d8383bd9c05073fa7187b71d2d834a204a/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java#L154) before `super.merge` is called. I'll try out that approach later today.